### PR TITLE
Replaces DNS TTL primitive types with Duration

### PR DESCRIPTION
**Pull request by ignasi35**
See https://github.com/akka/akka/pull/25848#discussion_r230058165

Must be merged after https://github.com/akka/akka/pull/25848

### CODE REVIEW

File: akka-actor/src/main/scala/akka/io/InetAddressDnsResolver.scala [hunk 1]
```diff
@@ -72,12 +73,12 @@ class InetAddressDnsResolver(cache: SimpleDnsCache, config: Config) extends Acto
         case None ⇒
           try {
             val answer = Dns.Resolved(name, InetAddress.getAllByName(name))
-            if (positiveTtl != Never) cache.put(answer, positiveTtl)
+            if (positiveTtl != Never) cache.put(answer, Duration.create(positiveTtl, TimeUnit.MILLISECONDS))
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/io/InetAddressDnsResolver.scala]:
```
import scala.concurrent.duration._

positiveTtl.millis
```

Comment (patriknw) [reply akka-actor/src/main/scala/akka/io/InetAddressDnsResolver.scala]: better is to do it once, in `getTtl` so that `val positiveTtl: FiniteDuration`
same with negativeTtl

File: akka-actor/src/main/scala/akka/io/SimpleDnsCache.scala [hunk 2]
```diff
@@ -64,8 +65,8 @@ object SimpleDnsCache {
       } yield e.answer
     }
 
-    def put(name: K, answer: V, ttlMillis: Long): Cache[K, V] = {
-      val until0 = clock() + ttlMillis
+    def put(name: K, answer: V, ttl: Duration): Cache[K, V] = {
+      val until0 = clock() + ttl.toMillis
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/io/SimpleDnsCache.scala]: This will throw `IllegalArgumentException` if `Duration` isn't a `FiniteDuration`, e.g. `Inf`

I think we should use `ttl: FiniteDuration` here.

File: akka-actor/src/main/scala/akka/io/dns/DnsResourceRecords.scala [hunk 3]
```diff
@@ -29,7 +31,7 @@ sealed abstract class ResourceRecord(val name: String, val ttl: Int, val recType
 }
 
 @ApiMayChange
-final case class ARecord(override val name: String, override val ttl: Int,
+final case class ARecord(override val name: String, override val ttl: Duration,
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/io/dns/DnsResourceRecords.scala]: Would need Java API as well, i.e. 
```
import akka.util.JavaDurationConverters._
def getTtl: java.time.Duration = ttl.asJava
```

Using Scala's `Duration` instead of `FiniteDuration` would complicate the Java API since there is no such thing as `Inf`. Perhaps easiest to use `FiniteDuration` with `Zero` as Never and `FiniteDuration(Long.MaxValue)` as Forever

File: akka-actor/src/main/scala/akka/io/dns/DnsResourceRecords.scala [hunk 4]
```diff
@@ -188,7 +190,7 @@ private[dns] object ResourceRecord {
     val name = DomainName.parse(it, msg)
     val recType = it.getShort
     val recClass = it.getShort
-    val ttl = it.getInt
+    val ttl = Duration.create(it.getInt, TimeUnit.SECONDS)
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/io/dns/DnsResourceRecords.scala]: it.getInt.seconds

**Comment by dwijnand** [comment]
You can rebase this now.

File: akka-actor/src/main/scala/akka/io/dns/DnsResourceRecords.scala (hunk 3 above)
Comment (ignasi35) [reply akka-actor/src/main/scala/akka/io/dns/DnsResourceRecords.scala]: I think modeling `Never` and `Forever` is the problem here. The code currently uses:

```scala
  // default values (-1 and 0 are magic numbers, trust them)
  private final val Forever = -1
  private final val Never = 0
  private final val DefaultPositive = 30
```
So the arithmetic _just works_ but there's a few `if`'s here and there making the code a bit had to follow. I'm considering using:

```scala
sealed trait CacheDuration{
  val toSeconds:Long
}
case object NeverCache extends CacheDuration {
  override val toSeconds: Long = 0L // for backwards compat
}
case object CacheForever extends CacheDuration {
  override val toSeconds: Long = -1L // for backwards compat
}
case class FiniteCache(ttl:FiniteDuration) extends CacheDuration {
  require(ttl.toSeconds>0 && ttl < Duration.Inf)
  override val toSeconds: Long = ttl.toSeconds
}

object CacheDuration{
  def fromInt(n:Int):CacheDuration={ n match{
    case 0 => CacheForever
    case n if n<0 => NeverCache
    case n => FiniteCache(n.seconds)
   }
  }
}
```
And move away from `Long` or `Duration/FiniteDuration`.

Comment (dwijnand) [reply akka-actor/src/main/scala/akka/io/dns/DnsResourceRecords.scala]: Sounds good, but I'd say:

* `CachePolicy` over `CacheDuration`
* I don't think we need the backwards compatibility (I would rather avoid it, especially with a name like `toSeconds`).  Or if we really need it maybe only in `SimpleDnsCache.Cache` itself (which is `private[io]`).

File: akka-actor-tests/src/test/scala/akka/io/InetAddressDnsResolverSpec.scala [hunk 5]
```diff
@@ -102,3 +102,31 @@ class InetAddressDnsResolverSpec extends AkkaSpec("""
...
+
+  private def dnsResolver = {
+    val actorRef = TestActorRef[InetAddressDnsResolver](Props(
+      classOf[InetAddressDnsResolver],
+      new SimpleDnsCache(),
+      system.settings.config.getConfig("akka.io.dns.inet-address")
+    ))
+    actorRef.underlyingActor
+  }
+}
```
Comment (ignasi35) [inline akka-actor-tests/src/test/scala/akka/io/InetAddressDnsResolverSpec.scala]: I noticed this test was missing while reviewing a regression I introduced.

File: akka-actor/src/main/scala/akka/io/InetAddressDnsResolver.scala [hunk 6]
```diff
@@ -29,41 +29,61 @@ class InetAddressDnsResolver(cache: SimpleDnsCache, config: Config) extends Acto
   // Deprecated JVM property key, keeping for legacy compatibility; replaced by NegativeCachePolicyProp
   private final val NegativeCachePolicyPropFallback = "sun.net.inetaddr.negative.ttl"
 
-  // default values (-1 and 0 are magic numbers, trust them)
-  private final val Forever = -1
-  private final val Never = 0
-  private final val DefaultPositive = 30
+  private final val DefaultPositive = FiniteCache(30.seconds)
```
Comment (ignasi35) [inline akka-actor/src/main/scala/akka/io/InetAddressDnsResolver.scala]: Remove these constants in favor of `Cachepolicy` ADT.

File: akka-actor/src/main/scala/akka/io/SimpleDnsCache.scala [hunk 7]
```diff
@@ -112,3 +116,23 @@ object SimpleDnsCache {
...
 }
+
+sealed trait CachePolicy
+case object NeverCache extends CachePolicy
+case object CacheForever extends CachePolicy
+case class FiniteCache(ttl: FiniteDuration) extends CachePolicy {
+  require(ttl.toSeconds > 0)
+  import akka.util.JavaDurationConverters._
+  def getTtl: java.time.Duration = ttl.asJava
+}
```
Comment (ignasi35) [inline akka-actor/src/main/scala/akka/io/SimpleDnsCache.scala]: This is the ADT.

File: akka-actor/src/main/scala/akka/io/SimpleDnsCache.scala [hunk 8]
```diff
@@ -112,3 +116,23 @@ object SimpleDnsCache {
...
+object FiniteCache {
+  // There's places where only a FiniteCache makes sense (DNS RFC says TTL is a positive 32 but integer)
+  // but we know the value can be cached effectively forever (e.g. the Lookup was the actual IP already)
+  val effectivelyForever: FiniteCache = FiniteCache(Int.MaxValue.seconds)
```
Comment (ignasi35) [inline akka-actor/src/main/scala/akka/io/SimpleDnsCache.scala]: Turns out I needed a `FiniteCache.effectivelyForever` which is typed `FiniteCache` but uses a value big enough.

The alternative was typing the `ttl` in all `DnsRecord` as `CachePolicy` and spread `asInstanceOf` in a few places (e.g. the `Ordering` type class).

File: akka-actor/src/main/scala/akka/io/dns/DnsResourceRecords.scala [hunk 9]
```diff
@@ -8,36 +8,39 @@ import java.net.{ Inet4Address, Inet6Address, InetAddress }
...
 import akka.io.dns.internal.{ DomainName, _ }
-import akka.util.{ ByteIterator, ByteString, ByteStringBuilder }
+import akka.util.{ ByteIterator, ByteString }
 
 import scala.annotation.switch
+import scala.concurrent.duration._
 
 @ApiMayChange
-sealed abstract class ResourceRecord(val name: String, val ttlInSeconds: Int, val recType: Short, val recClass: Short)
+sealed abstract class ResourceRecord(val name: String, val ttl: FiniteCache, val recType: Short, val recClass: Short)
```
Comment (ignasi35) [inline akka-actor/src/main/scala/akka/io/dns/DnsResourceRecords.scala]: DNS RFC states the TTL is a positive 32 bit integer so the appropriate type here is `FiniteCache` since `NeverCache` and `CacheForever` are forbidden.

File: akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsCache.scala [hunk 10]
```diff
@@ -16,12 +16,13 @@ import akka.io.dns.internal.DnsClient.Answer
...
 
 import scala.annotation.tailrec
+import scala.concurrent.duration._
 
 /**
  * Internal API
  */
 @InternalApi class AsyncDnsCache extends Dns with PeriodicCacheCleanup {
-  private val cache = new AtomicReference(new Cache[(String, QueryType), Answer](
+  private val cacheRef = new AtomicReference(new Cache[(String, QueryType), Answer](
```
Comment (ignasi35) [inline akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsCache.scala]: Renamed to avoid confusion.

File: akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala [hunk 11]
```diff
@@ -60,8 +59,8 @@ private[io] final class AsyncDnsResolver(
         Try {
           val address = InetAddress.getByName(name) // only checks validity, since known to be IP address
           val record = address match {
-            case _: Inet4Address           ⇒ ARecord(name, Int.MaxValue, address)
-            case ipv6address: Inet6Address ⇒ AAAARecord(name, Int.MaxValue, ipv6address)
+            case _: Inet4Address           ⇒ ARecord(name, FiniteCache.effectivelyForever, address)
+            case ipv6address: Inet6Address ⇒ AAAARecord(name, FiniteCache.effectivelyForever, ipv6address)
```
Comment (ignasi35) [inline akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala]: When the `name` in the `Lookup` is a valid IP, the query is short-circuited. This use case requires `FiniteCache.effectivelyForever`.

File: akka-actor/src/main/scala/akka/io/InetAddressDnsResolver.scala [hunk 12]
```diff
@@ -29,41 +29,61 @@ class InetAddressDnsResolver(cache: SimpleDnsCache, config: Config) extends Acto
...
       .orElse(Try(System.getProperty(NegativeCachePolicyPropFallback).toInt))
-      .getOrElse(0) // default
-    if (n < 0) Forever else n
+      .map(parsePolicy)
+      .getOrElse(NeverCache) // default
+
+  private def parsePolicy(n: Int): CachePolicy = {
+    n match {
+      case 0          ⇒ CacheForever
+      case x if x < 0 ⇒ NeverCache
```
Comment (dwijnand) [inline akka-actor/src/main/scala/akka/io/InetAddressDnsResolver.scala]: This is the opposite in than the original.

In the original (above) `n < 0` matches `CacheForever` (i.e `-1`) and `0` matches `NeverCache`.  Why the change?

**Review by dwijnand — COMMENTED** [review]
This is great.

There's one part the concerns me, which I've commented.

Other than that, I wanted to say that I think the name `effectivelyForever` is great.  I'm still not loving the name `FiniteCache`, on the other hand...  What do you think about calling it `CacheDuration` instead?

File: akka-actor/src/main/scala/akka/io/InetAddressDnsResolver.scala (hunk 12 above)
Comment (ignasi35) [reply akka-actor/src/main/scala/akka/io/InetAddressDnsResolver.scala]: It's a mistake. Another missing test, I guess.

Comment (ignasi35) [reply akka-actor/src/main/scala/akka/io/InetAddressDnsResolver.scala]: Thanks @dwijnand

**Comment by ignasi35** [comment]
>  I'm still not loving the name FiniteCache, on the other hand... What do you think about calling it CacheDuration instead?

Agree, `FiniteCache` is not a great name. I was actually thinking maybe `Ttl` could work. Imagine the following snippets:

```scala
private def parsePolicy(n: Int): CachePolicy = {
  n match {
    case 0          ⇒ NeverCache
    case x if x < 0 ⇒ CacheForever
    case x          ⇒ FiniteCache/CacheDuration/Ttl(x.seconds)
  }
}
// -------
class ResourceRecord(..., val ttl: FiniteCache/CacheDuration/Ttl, ...)
```
WDYT?

**Comment by dwijnand** [comment]
`Ttl` works for me!

**Comment by dwijnand** [comment]
Btw with `Ttl` I would call accessing the inner value `value`/`.getValue` so you end up with `ttl.value/ttl.getValue` instead of `ttl.ttl/ttl.getTtl`.

**Comment by hepin1989** [comment]
Why not just using works never and infinite for NeverCache and CacheForever.

File: akka-actor/src/main/scala/akka/io/InetAddressDnsResolver.scala (hunk 12 above)
Comment (hepin1989) [reply akka-actor/src/main/scala/akka/io/InetAddressDnsResolver.scala]: case "infinite" => CacheForever
case "never" => NeverCache

**Comment by ignasi35** [comment]
Thanks @hepin1989 and @dwijnand for the naming recommendations. :-)

**Comment by ignasi35** [comment]
rebased in master

**Comment by ignasi35** [comment]
I think `SimpleDnsCache` and `CachePolicy` could be moved into the `akka.io.dns` package.

**Review by dwijnand — APPROVED** [review]
Mentioned to Ignasi in another conversation I was having with him that I think `Never` and `Forever` are a bit too generic-sounding names to be top-level types (even if residing in the `akka.io` package).

But it's a minor concern. LGTM

File: akka-actor/src/main/scala/akka/io/SimpleDnsCache.scala [hunk 13]
```diff
@@ -36,10 +37,10 @@ class SimpleDnsCache extends Dns with PeriodicCacheCleanup {
   }
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/io/SimpleDnsCache.scala]: `class SimpleDnsCache` can't be moved to another package if that was the question.
It's public api (not sure why it's public but that is how it is).

**Review by patriknw — COMMENTED** [review]
looks like a good approach

**Comment by dwijnand** [comment]
I don't think we can make them internal API.

`Ttl` is the type of one of the elements of the public, `@ApiMayChange` record types in `akka.io.dns`.

Additionally, `CachePolicy` is the return type of `positiveCachePolicy`/`negativeCachePolicy` in public `InetAddressDnsResolver`.

**Comment by ignasi35** [comment]
> I don't think we can make them internal API.

Good call. It looks like `CachePolicy` _belongs_ to `akka.io.dns`. I'll mark as `ApiMayChange`, though.

**Comment by patriknw** [comment]
@ignasi35 I think this is looking good. Is it ready for final review? (then remove WIP from title)

**Comment by ignasi35** [comment]
Thanks @patriknw, I think it's ready, yeah. I'm reviewing MiMa.

File: akka-actor/src/main/scala/akka/io/InetAddressDnsResolver.scala [hunk 14]
```diff
@@ -29,41 +30,55 @@ class InetAddressDnsResolver(cache: SimpleDnsCache, config: Config) extends Acto
...
+      case "default" ⇒ if (positive) defaultCachePolicy else defaultNegativeCachePolicy
+      case "forever" ⇒ Forever
       case "never"   ⇒ Never
-      case _ ⇒ config.getDuration(path, TimeUnit.MILLISECONDS)
-        .requiring(_ > 0, s"akka.io.dns.$path must be 'default', 'forever', 'never' or positive duration")
+      case _ ⇒ {
+        val finiteTtl = config
+          .getDuration(path, TimeUnit.MILLISECONDS)
+          .requiring(_ > 0, s"akka.io.dns.$path must be 'default', 'forever', 'never' or positive duration")
+        Ttl.fromPositive(finiteTtl.seconds)
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/io/InetAddressDnsResolver.scala]: isn't this strange; getDuration in MILLISECONDS, and here `.seconds` ?

File: akka-actor/src/main/scala/akka/io/InetAddressDnsResolver.scala [hunk 15]
```diff
@@ -29,41 +30,55 @@ class InetAddressDnsResolver(cache: SimpleDnsCache, config: Config) extends Acto
...
-  private final val DefaultPositive = 30
+  private final val DefaultPositive = Ttl.fromPositive(30.seconds)
 
-  private lazy val cachePolicy: Int = {
-    val n = Try(Security.getProperty(CachePolicyProp).toInt)
+  private lazy val defaultCachePolicy: CachePolicy =
+    Try(Security.getProperty(CachePolicyProp).toInt)
       .orElse(Try(System.getProperty(CachePolicyPropFallback).toInt))
+      .map(parsePolicy)
       .getOrElse(DefaultPositive) // default
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/io/InetAddressDnsResolver.scala]: if these properties are misconfigured, e.g. not a number, then we are silently using the defaults? logging?

File: akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+object CachePolicy {
+
+  @ApiMayChange
+  sealed trait CachePolicy
+  @ApiMayChange
+  case object Never extends CachePolicy
+  @ApiMayChange
+  case object Forever extends CachePolicy
+  @ApiMayChange
+  sealed abstract case class Ttl(value: FiniteDuration) extends CachePolicy {
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala]: why is this `abstract` (case classes should be final in my opinion)
if you want to enforce creation from the factory methods in the companion you can make the constructor private `final case class Ttl private (value: FiniteDuration)`

Comment (ignasi35) [reply akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala]: IIRC, making this `abstract` prevents the creation of `copy` and makes it impossible to break the invariant we want to enforce (`value > Duration.zero`) which is on the `fromPositive` factory method.

Comment (patriknw) [reply akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala]: doesn't prevent new instances to be created from java

that invariant should be checked in the constructor instead

Comment (dwijnand) [reply akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala]: :+1: for making the constructor private (good catch)
:-1: for throwing in the constructor: throw in the obviously-named factory method and disallow the copy method

Comment (dwijnand) [reply akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala]: (more info: https://twitter.com/tpolecat/status/788215281046917120)

File: akka-actor/src/main/scala/akka/io/InetAddressDnsResolver.scala (hunk 15 above)
Comment (ignasi35) [reply akka-actor/src/main/scala/akka/io/InetAddressDnsResolver.scala]: Hmmm, good idea. 

The class currently doesn't have a logger injected so I'm tempted to add a good old `LoggerFactory.getLogger` but I'm not sure that's good idea in the Akka codebase, though.

File: akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala (hunk 16 above)
Comment (patriknw) [reply akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala]: private constructor would not be private from java (I think) so if we really care about the invariant it should be in the constructor

the obviously-named factory method will be in the stack trace

File: akka-actor/src/main/scala/akka/io/InetAddressDnsResolver.scala (hunk 15 above)
Comment (ignasi35) [reply akka-actor/src/main/scala/akka/io/InetAddressDnsResolver.scala]: The alternative is a new constructor argument (`logger: LoggingAdapter`) but this is public APi (not even `@ApiMayChange`). :-(

Comment (patriknw) [reply akka-actor/src/main/scala/akka/io/InetAddressDnsResolver.scala]:
```
extends Actor with ActorLogging
```

then you have `log.`

File: akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala (hunk 16 above)
Comment (dwijnand) [reply akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala]: Ah yes you're right!  In order to allow the companion object to access it it can't be private.  It could be java package private, not sure why it isn't...

So move the check into the constructor
Just call new in the factory method
Keep abstract to elide generating copy.

Comment (dwijnand) [reply akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala]: Using `abstract` is trivially easier to add than re-implementing unapply, equals, hashCode and toString.  😄

But you're right that you can't maintain `unapply`'s binary compatibility with `case class`...

File: akka-actor/src/main/scala/akka/io/InetAddressDnsResolver.scala (hunk 15 above)
Comment (ignasi35) [reply akka-actor/src/main/scala/akka/io/InetAddressDnsResolver.scala]: 🤦‍♂️ ofc

File: akka-actor/src/main/scala/akka/io/dns/DnsResourceRecords.scala (hunk 9 above)
Comment (ignasi35) [reply akka-actor/src/main/scala/akka/io/dns/DnsResourceRecords.scala]: BTW, here's the link: https://www.ietf.org/rfc/rfc1035.txt

File: akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala (hunk 16 above)
Comment (ignasi35) [reply akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala]: Thanks again @patriknw! We overlooked the java use cases in earlier iterations. I've just pushed a version that I think reflects all the suggested changes and includes `apply` as a bonus.

**Comment by ignasi35** [comment]
Rebased in `master` and fixed (?) MiMa.

File: akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+object CachePolicy {
+
+  @ApiMayChange
+  sealed trait CachePolicy
+  @ApiMayChange
+  case object Never extends CachePolicy
+  @ApiMayChange
+  case object Forever extends CachePolicy
+  @ApiMayChange
+  class Ttl(val value: FiniteDuration) extends CachePolicy {
```
Comment (dwijnand) [inline akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala]:
```suggestion
  final class Ttl(val value: FiniteDuration) extends CachePolicy {
```

File: akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+    if (value <= Duration.Zero) throw new IllegalArgumentException(s"TTL values must be a positive, 32-bit integer.")
+    import akka.util.JavaDurationConverters._
+    def getValue: java.time.Duration = value.asJava
+
+    def canEqual(other: Any): Boolean = other.isInstanceOf[Ttl]
+
+    override def equals(other: Any): Boolean = other match {
+      case that: Ttl ⇒
+        (that canEqual this) &&
```
Comment (dwijnand) [inline akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala]: This is always true.  There's no need to define and use `Objects.hash(what)` with `Ttl` being final (which it should be).

File: akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+          value == that.value
+      case _ ⇒ false
+    }
+
+    override def hashCode(): Int = {
+      val state = Seq(value)
+      state.map(_.hashCode()).foldLeft(0)((a, b) ⇒ 31 * a + b)
```
Comment (dwijnand) [inline akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala]: Could be simpler...

```suggestion
      java.util.Objects.hash(value)
```

File: akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    override def toString = s"Ttl($value)"
+  }
+  @ApiMayChange
+  object Ttl {
+    def unapply(ttl: Ttl): Option[FiniteDuration] = Some(ttl.value)
+    def apply(value: FiniteDuration): Ttl = {
```
Comment (dwijnand) [inline akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala]: With the constructor throwing this should be `fromPositive` as (IMO) `apply` implies a pure function.

**Review by dwijnand — CHANGES_REQUESTED** [review]
Left a few notes.  The most important ones are the `apply` and the `final`.

File: akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala (hunk 19 above)
Comment (ignasi35) [reply akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala]: TIL

**Review by dwijnand — APPROVED** [review]
LGTM! 👍

File: akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala [hunk 21]
```diff
@@ -0,0 +1,52 @@
...
+object CachePolicy {
+
+  @ApiMayChange
+  sealed trait CachePolicy
+  @ApiMayChange
+  case object Never extends CachePolicy
+  @ApiMayChange
+  case object Forever extends CachePolicy
+  @ApiMayChange
+  final class Ttl(val value: FiniteDuration) extends CachePolicy {
```
Comment (dwijnand) [inline akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala]: The constructor should be (Scala) `private`:
```suggestion
  final class Ttl private (val value: FiniteDuration) extends CachePolicy {
```

File: akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala [hunk 22]
```diff
@@ -0,0 +1,52 @@
...
+    if (value <= Duration.Zero) throw new IllegalArgumentException(s"TTL values must be a positive, 32-bit integer.")
```
[the remark this answers is not in the document]
Comment (ignasi35) [reply akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala]: 😅the message was inspired by the [RFC](https://www.ietf.org/rfc/rfc1035.txt). Out of context is misleading.

File: akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala [hunk 23]
```diff
@@ -0,0 +1,52 @@
...
+    import akka.util.JavaDurationConverters._
+    def getValue: java.time.Duration = value.asJava
+
+    override def equals(other: Any): Boolean = other match {
+      case that: Ttl ⇒ value == that.value
+      case _         ⇒ false
+    }
+
+    override def hashCode(): Int = {
+      java.util.Objects.hash(value)
```
[the remark this answers is not in the document]
Comment (ignasi35) [reply akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala]: Fixing.

**Comment by ignasi35** [comment]
I think this is ready to be merged.

**Review by patriknw — APPROVED** [review]
LGTM

**Comment by patriknw** [comment]
@chbatey I think this has been thoroughly reviewed by several people but since you have been into most of the DNS stuff you could take a brief final look before merging.

**Comment by patriknw** [comment]
Refs https://github.com/akka/akka/issues/26015
Important to merge this before releasing 2.5.19 since it's a breaking API change and we don't want to break it twice.

File: akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala [hunk 24]
```diff
@@ -0,0 +1,50 @@
...
+object CachePolicy {
+
+  @ApiMayChange
+  sealed trait CachePolicy
+  @ApiMayChange
+  case object Never extends CachePolicy
+  @ApiMayChange
+  case object Forever extends CachePolicy
+  @ApiMayChange
+  final class Ttl private (val value: FiniteDuration) extends CachePolicy {
```
Comment (hepin1989) [inline akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala]: why not `TTL`?

File: akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala [hunk 25]
```diff
@@ -0,0 +1,50 @@
...
+    override def hashCode(): Int = value.hashCode()
+
+    override def toString = s"Ttl($value)"
+  }
+  @ApiMayChange
+  object Ttl {
+    def unapply(ttl: Ttl): Option[FiniteDuration] = Some(ttl.value)
+    def fromPositive(value: FiniteDuration): Ttl = {
+      new Ttl(value)
+    }
```
Comment (hepin1989) [inline akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala]: Where is the Java API?

Comment (dwijnand) [reply akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala]: Well spotted, @hepin1989.

@ignasi35 add an overload which takes `java.time.Duration` and delegate to the scala impl by calling `.asScala` after file-wide importing `import akka.util.JavaDurationConverters._`.

File: akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala (hunk 24 above)
Comment (dwijnand) [reply akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala]: `TTL` isn't terrible, but I prefer this style, (which is nicely defined in the Google style guide: https://google.github.io/styleguide/javaguide.html#s5.3-camel-case)

File: akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala (hunk 25 above)
Comment (dwijnand) [reply akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala]: https://github.com/ignasi35/akka/pull/1

File: akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala (hunk 24 above)
Comment (hepin1989) [reply akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala]: Both ok then.

File: akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala (hunk 25 above)
Comment (ignasi35) [reply akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala]: Thanks for the callout @hepin1989 ! :-)
And thanks @dwijnand for the PR

**Comment by dwijnand** [comment]
Ah, damn, now the PR is mixed authors... 🙄

You squashers.

File: akka-actor-tests/src/test/scala/akka/io/InetAddressDnsResolverSpec.scala (hunk 5 above)
Comment (chbatey) [reply akka-actor-tests/src/test/scala/akka/io/InetAddressDnsResolverSpec.scala]: :+1:

**Review by chbatey — APPROVED** [review]
LGTM

**Comment by dwijnand** [comment]
@nvollmar good question.  The logging came from this comment: https://github.com/akka/akka/pull/25850#discussion_r233055997, but it should've been applied to the more tighter scope of misconfiguration.

**Comment by dwijnand** [comment]
https://github.com/akka/akka/pull/26249
